share/doc/iso/wiscman/cons.4
.\"
.\" 5799-WZQ (C) COPYRIGHT IBM CORPORATION 1986,1987,1988
.\" LICENSED MATERIALS - PROPERTY OF IBM
.\" REFER TO COPYRIGHT INSTRUCTIONS FORM NUMBER G120-2083
.\"
.\"$Header: /usr/tmp/cvs2git/cvsroot-netbsd/src/share/doc/iso/wiscman/Attic/cons.4,v 1.1 1994/06/19 00:07:22 cgd Exp $
.\"$ACIS:cons.4_ca 11.3$
.\"$Source: /usr/tmp/cvs2git/cvsroot-netbsd/src/share/doc/iso/wiscman/Attic/cons.4,v $
.\" This file uses -man macros.
.TH CONS 4 "Sept 1988" "Space overwritten by .AC macro" " " 
.AC 1 0 
.SH NAME
cons \- keyboard and console display interface
.SH DESCRIPTION
The keyboard and various possible displays combine to 
provide a terminal-like
interface to the system. Internally, these are separate devices which
software combines to emulate a normal terminal. See the appropriate manual
pages for information about each display and the keyboard.
.PP
The keyboard adapter also supports the speaker, which is activated 
when the ASCII character \fBbel\fP (\fB^G\fP)
is sent to the display with software.
For additional information on speaker control, see \fIspeaker\fP(4).
.PP
.B Console Device Control
.PP
The display devices, 
\fI/dev/ttyaed\fR, \fI/dev/ttyap16\fR, \fI/dev/ttyap8c\fR,
\fI/dev/ttyapa8\fR, \fI/dev/ttyega\fR, \fI/dev/ttymono\fR, 
\fI/dev/ttympel\fR, \fI/dev/ttyvga\fR, and \fI/dev/tty8514\fR are all
minor devices under 
\fI/dev/console\fR, and are all capable of displaying console output. 
Unique to this system is the fact that you may have one or more of these
displays on your workstation at a time and any one can act as a console.
With only one keyboard and system mouse, the console driver
multiplexes these input devices to the many displays.
All of the displays may have simultaneous logins and the user
can ``hot key'' between each display.
At first, this 
``input focus''
is on
the first device in the above sequence to 
be found at initialization time.  The input focus
can be manually switched to the next available display by pressing the
default ``hot key'' <Alt><Scroll Lock>. 
When the
input focus
is on a display, all keyboard and mouse data are sent to the process(es)
that read from that display.
.PP
If no other console tty device is open, and only the default input 
emulator is used (see \fIkbdemul\fP(4)), the input focus is set to
\fI/dev/console\fP.  In this case, <Alt><Scroll Lock> only switches
which display gets console output.
In the case where one or more tty devices are open, or the default input
emulator changes, \fI/dev/console\fP gets no input.  It tries to send output
to the currently focused device.  A user can redirect these console messages
to any tty devices with the TIOCCONS ioctl.
.PP
To support the many displays and the multiplexing between them, an
emulator package was developed to work with the console driver.
This package allows different types of emulation on input and output to
be written independently of device.
.PP
The display devices \fI/dev/aed\fP, \fI/dev/apa16\fP, \fI/dev/apa8c\fP,
\fI/dev/apa8\fP, \fI/dev/ega\fR, \fI/dev/mono\fP, 
\fI/dev/mpel\fR, \fI/dev/vga\fR, and \fI/dev/ibm8514\fR 
are also minor devices to 
\fI/dev/console\fP.  They are typically used by window managers and other
graphic applications.  When the focus is pointed to one of these display
devices, the console messages are put in a circular buffer 
(see \fIbufemul\fP(4))
unless redirected with the TIOCCONS ioctl.
The buffer is flushed to the screen upon closing the display device.
.PP
The following are generic console \fIioctls\fP defined in \fIscreen_cousf.u\fP:
.TP 20
CON_SELECT_SCREEN 
Output focus is set to display number (arg > 0) or 
to next display in list (arg < 0).  Previous display number is returned.  
.TP 20
CON_GET_SCREEN 
Just returns the current output focus display number.  
.TP 20
EIGETD 
Gets the number of the current input emulator for this display.  
.TP 20
EOGETD 
Gets the number of the current output emulator for this display.  
.TP 20
EISETD 
Sets the input emulator and returns the previous for this display.  
.TP 20
EOSETD 
Sets the output emulator and returns the previous for this display.  
.TP 20
CON_INIT_SCREEN 
Initializes the specified display (arg >=0) or this display
(arg < 0).  
.TP 20
CON_GET_FOCUS_CODE 
Gets the current keyboard code for setting the console 
focus (affects xemul only).  
.TP 20
CON_SET_FOCUS_CODE 
Sets the current keyboard code for setting the console
focus (affects xemul only), and return the previous code.
.PP
All of the above commands take integer arguments.  
.PP
The following are generic console \fIioctls\fP defined in \fIconsio.h\fP: 
.TP 20
SCRIOCGETF 
Gets the screen control flags for the given display number.  
.TP 20
SCRIOCSETC 
Sets the screen control flags for the given display number.  
.PP
.in +10
SCRIOCGETF and SCRIOCSETC use the following structure:
.DT
.nf
struct screen_control { 
			 int   device; /* which screen/display to control */  
			 int   switches; /* Flags for this screen */
};
.fi
.sp 2
.RS 6
.TP 20
CONSDEV_PRESENT 
Display is present on this system.  
This bit cannot be changed by SCRIOSETC. 
.TP 20
CONSDEV_KERNEL 
Display is available to the kernel.  
.TP 20
CONSDEV_USER 
Display is available to the user.  
.TP 20
CONSDEV_INIT 
Display has been initialized for output.  
.TP 20
CONSDEV_TTY 
Diplay has been initialuzed for output.  
This bit cannot be changed by SCRIOSETC.
.TP 20
CONSDEV_GRA 
Graphics display has been opened directly by minor device number. 
This bit cannot be changed by SCRIOSETC.
.TP 20
CONSDEV_NOINPUT 
Prevents the "round-robin" console focus-switching from finding this 
display.  This flag is cleared when \fIbuf_emul\fP is closed.  
.TP 20
SCRSETNIP 
Sets the no-input bit in the screen control flags for the 
display's current file description.  
.TP 20
SCRCLRNIP 
Clears the no-input bit in the screen control flags for the 
display's current file description.  
.RE
.PP
The following \fIioctl\fP is defined in \fIbufemul.h\fP: 
.TP 20
BUFDISPINFO 
\fIArg\fP returns the following information about the display: 
.PP
.RS 6
.TP 20
BUF_IS_ATR(arg) 
True when the CPU is an IBM 6152 Academic System.  
.TP 20
BUF_IS_RTPC(arg) 
True when the CPU is an IBM RT PC.  
.TP 20
BUF_GET_PCCODE(x) 
Get PC-Code version/type byte (IBM 6152 only).  
.bp
.TP 20
BUF_GET_VGA(arg) 
Get the type of display connected to the VGA.  
0=none, 1=color, 2=gray.  Valid only for the IBM 6152 Academic System.   
.TP 20
BUF_GET_8514(arg) 
Get the type of display connected to the IBM 8514/A.  
0=none, 1=color, 2=gray.  Valid only for the IBM 6152 Academic System.  
.TP 20
BUF_GET_EGA(arg) 
Returns the value of the switches on the EGA display.  Valid only for the IBM RT PC with an EGA card installed.  
.RE
.PP
All of the above \fIioctl\fP system calls are device-independent controls 
for dealing with the emulators.  
.PP
Each emulator has its own set of \fIioctls\fP for its own emulation purposes.  
These other \fIiotls\fP are used in window-manager emulators for operations 
such as passing/positioning the mouse locator for/on the display.  
See the man page for any particular emulator for more information.  
.PP
.SH NOTE
On the IBM RT PC, the kernel flashes ``98'' on the LEDs if it cannot find any
configured display during initialization, and then proceeds. 
.SH DIAGNOSTICS
None.
.SH FILES
.PP
For the IBM RT PC:
.br
/dev/console 
.br
/dev/aed 
.br
/dev/apa16
.br
/dev/apa8c
.br
/dev/apa8
.br
/dev/ega
.br
/dev/mono
.br
/dev/mpel
.br
.PP
For the IBM 6152 Academic System:
.br
/dev/vga
.br
/dev/ibm8514
.SH "SEE ALSO"
bufemul(4), bus(4), ibm5081(4), ibm5151(4), ibm6153(4), ibm6154(4),
ibm6155(4), ibm8514(4), ibmaed(4), ibmemul(4), kbdemul(4), 
speaker(4), stdemul(4), tty(4), vga(4), xemul(4), setscreen(8)
.br
``IBM/4.3 Console Emulators'', in Volume II, Supplementary Documents




doc/CHANGES
# LIST OF CHANGES FROM LAST RELEASE:			<$Revision: 1.3291 $>
#
#
# [Note: This file does not mention every change made to the NetBSD source tree.
# For an exhaustive list, please subscribe to the `source-changes' mailing list,
# or see the mailing list archives.  For more information, send a message
# containing just the word `help' to `majordomo@NetBSD.org'.]
#
# ----------------------------------------------------------------------------
# PLEASE USE THE FOLLOWING STYLE WHEN ADDING ITEMS:
#
#	theme: Content ...
#		more content ...
#		more content. [developer DATE-FORMAT]
#
# That's it, one tab to start an item, and two tabs to indent the content.
# Also please start the content with a capital and end it with a dot.
#
# This is very important for the www scripts, otherwise it won't be able to
# parse the file and the HTML output will be mangled.
#
# See htutils/changes/changes2html script for more details.
#
# ----------------------------------------------------------------------------

Changes from NetBSD 11.0 to NetBSD 12.0:

	macppc: Add drivebay(4), a driver to control the Xserve G4's drive bays
		via sysctl. [macallan 20250812]
	macppc: Added support for fan control on Xserve G4. [macallan 20250820]
	acpi(4): Updated ACPICA to 20250807. [christos 20250824]
	binutils: Import 2.45 and switch amd64 to it  [christos 20250826]
	gffb(4): Add backlight control to the GeForce framebuffer driver.
		[macallan 20250830]
	gdb: Import 16.3 and switch everyone to it [christos 20250902]
	openpam(3): update to 20250531 (zingiber) [christos 20250903]
	openldap: Import 2.6.10. [christos 20250904]
	unbound(8): Import 1.23.1. [christos 20250906]
	kernel: Add new memory barrier for paravirtualized drivers
		requiring store-before-load ordering. [riastradh 20250906]
	macppc: Add driver for the LED bar found on Xserve G4,
		controls lights depending on CPU usage. [macallan 20250908]
	m68k: Change softfloat build to libc/softfloat. [nat 20250910]
	fsck_ffs(8): Expand flags with -a and -i. [perseant 20250914]
	dumplfs(8): Add -n flag to allow dumping of a given inode from an
		inode block, or every inode in a block. [perseant 20250914]
	lfs: Add working in-kernel roll forward. [perseant 20250917]
	kernel: Add a shared function to query the common properties used
		for configuring an Ethernet address. [thorpej 20251004]
	gffb(4): Add backlight control for NVIDIA GeForce FX 5200 Go.
		[macallan 20251006]
	hypotl(3): Accommodate infinity for softfloat. [nat 20251007]
	arm, aarch64: Switch to GCC 14. [mrg 20251007]
	hppa: Switch to GCC 14. [mrg 20251007]
	riscv: Switch to GCC 14. [mrg 20251007]
	sparc, sparc64: Switch to GCC 14. [mrg 20251007]
	vax: Switch to GCC 14. [mrg 20251007]
	x86: Switch to GCC 14. [mrg 20251007]
	ucom(4): Add support for console over USB-to-serial adapter
		[manu 20251009]
	libc: Merge extended worker threads for POSIX asynchronous I/O
		(GSoC project by Ethan Miller). [christos 20251010]
	ssh(1), sshd(8): Import OpenSSH 10.2. [christos 20251011]
	gpt(8): Add various riscv-specific partition types. [thorpej 20251011]
	fsck_lfs(8): Enable checking LFS64 volumes. [perseant 20251012]
	m68k: New port: mac68ksf - Macintosh 68k softfloat. [nat 20251020]
	sc16is7xx(4): NXP SC16IX7xx family of UART chips that are behind
		a I2C or SPI bus.  [brad 20251024]
	compat_linux(8): Add support for semtimedop(2), from Ricardo Branco.
		[christos 20251026]
	ftp(1): Add "connect" as a synonym for "open". [lukem 20251028]
	ia64: Switch to GCC 14. [mrg 20251029]
	lfs: Add in-kernel cleaner. [perseant 20251106]
	m68k: Experimental new pmap module for systems with 68k MMUs
		(plus the HP MMU). [thorpej 20251108]
	compat_linux(8): Add prctl(2) lwp_{g,s}etname, from Kevin Bloom.
		[christos 20251110]
	sparc64: Add implementation of Sun's VIS instruction API.
		[nia 20251113]
	iconv(1): Support -o outfile like gnu does and check writes for
		failure. [christos 20251121]
	cgd(4): Improve throughput of AES disk encryption using SSE2.
		[riastradh 20251123]
	powerpc: Switch to GCC 14. [martin 20251128]
	dreamcast, atu(4), icp(4), gpsa(8): Relicense various NetBSD
		components from 4-clause to 3-clause BSD. [nia 20251128]
	umcpmio(4): Reorg and support for the MCP2210, USB to SPI
		bridge. [brad 20251129]
	ucom(4): Allow using as boot console. [manu 20251129]
	sparc64: Use 64-bit optimized AES code for disk encryption.
		[nia 20251201]
	iwm(4): Update driver code, improving performance and stability.
		[mlelstv 20251201]
	lfs: Various stability improvements. [perseant 20251202]
	viaide(4): Identify and attach VT6420 IDE (PATA) controller.
		[andvar 20251204]
	diff(1), diff3(1): Import BSD-licensed diff to replace GNU
		diffutils, build with MKBSDDIFF=yes. [nia 20251204]
	evbppc: Enable PaX security features in WII kernel configuration.
		[nia 20251204]
	amd64: Support ramdisks when booting in GENPVH mode.
		[khorben 20251205]
	ftp(1): Fix ASCII transfers with progress bar. [lukem 20251206]
	evbppc: wii: Add support for GameCube controller sockets
		as uhid(4) (now "userspace hid") devices.
		[jmcneill 20251208]
	mips: Switch to GCC 14. [mrg 20251209]
	openresolv(8): import 3.17.1 (patched to 3.17.3) [roy 20251213]
	editline(3): Add a method to set custom word characters,
		from bsiegert. [christos 20251214]
	evbarm: Add support for TI OMAP4 PandaBoard-ES, from Rui-Xiang Guo.
		[skrll 20251216]
	tzcode: Updated to 2025c. [christos 20251218]
	tzdata: Updated to 2025c (using 2025cgtz) [kre 20251219]
	re(4): Made the code MPSAFE with a private mutex.
		[mlelstv 20251220]
	lm(4): Support five aux fan sensors for NCT6794D. [isaki 20251224]
	inetd(8): Add an optional accept limit to stream/nowait services.
		[mlelstv 20251227]
	sun2: Add various MK build knobs and disable them on sun2 to trim
		down the installation. [thorpej 20251228]
	diff(1): Switch the default to MKBSDDIFF=yes. [nia 20260102]
	i386, amd64, sparc64: Exclude various base features from the
		CD-ROM installer and add checks that cause ISO image

sys/arch/mac68k/mac68k/cons.c
/*
 * Copyright (c) 1988 University of Utah.
 * Copyright (c) 1982, 1990 The Regents of the University of California.
 * All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * the Systems Programming Group of the University of Utah Computer
 * Science Department.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */
#ident "$Id: cons.c,v 1.1 1993/09/29 06:09:17 briggs Exp $"
/*
 * from: Utah $Hdr: cons.c 1.1 90/07/09$
 *
 *	@(#)cons.c	7.6 (Berkeley) 5/4/91
 */

#include "sys/param.h"
#include "sys/proc.h"
#include "sys/systm.h"
#include "sys/buf.h"
#include "sys/ioctl.h"
#include "sys/tty.h"
#include "sys/file.h"
#include "sys/conf.h"

#include "cons.h"

int	macprobe(), macinit(), macgetchar(), macputchar();
int	sercnprobe(), sercninit(), sercngetc(), sercnputc();

struct	consdev constab[] = {
	{ macprobe,	macinit,	macgetchar,	macputchar },
	/* { sercnprobe,	sercninit,	sercngetc,	sercnputc }, */
	{ 0 },
};

struct	tty *constty = 0;	/* virtual console output device */
struct	consdev *cn_tab;	/* physical console device info */
struct	tty *cn_tty;		/* XXX: console tty struct for tprintf */

cninit()
{
	register struct consdev *cp;

	/*
	 * Collect information about all possible consoles
	 * and find the one with highest priority
	 */
	for (cp = constab; cp->cn_probe; cp++) {
		(*cp->cn_probe)(cp);
		if (cp->cn_pri > CN_DEAD &&
		    (cn_tab == NULL || cp->cn_pri > cn_tab->cn_pri))
			cn_tab = cp;
	}
	/*
	 * No console, we can handle it
	 */
	if ((cp = cn_tab) == NULL)
		return;
	/*
	 * Turn on console
	 */
	cn_tty = cp->cn_tp;
	(*cp->cn_init)(cp);
}

cnopen(dev, flag, mode, p)
	dev_t dev;
	int flag, mode;
	struct proc *p;
{
	if (cn_tab == NULL)
		return (0);
	dev = cn_tab->cn_dev;
	return ((*cdevsw[major(dev)].d_open)(dev, flag, mode, p));
}
 
cnclose(dev, flag, mode, p)
	dev_t dev;
	int flag, mode;
	struct proc *p;
{
	if (cn_tab == NULL)
		return (0);
	dev = cn_tab->cn_dev;
	return ((*cdevsw[major(dev)].d_close)(dev, flag, mode, p));
}
 
cnread(dev, uio, flag)
	dev_t dev;
	struct uio *uio;
	int flag;
{
	if (cn_tab == NULL)
		return (0);
	dev = cn_tab->cn_dev;
	return ((*cdevsw[major(dev)].d_read)(dev, uio, flag));
}
 
cnwrite(dev, uio, flag)
	dev_t dev;
	struct uio *uio;
	int flag;
{
	if (cn_tab == NULL)
		return (0);
	dev = cn_tab->cn_dev;
	return ((*cdevsw[major(dev)].d_write)(dev, uio, flag));
}
 
cnioctl(dev, cmd, data, flag, p)
	dev_t dev;
	int cmd;
	caddr_t data;
	int flag;
	struct proc *p;
{
	int error;

	if (cn_tab == NULL)
		return (0);
	/*
	 * Superuser can always use this to wrest control of console
	 * output from the "virtual" console.
	 */
	if (cmd == TIOCCONS && constty) {
		error = suser(p->p_ucred, (u_short *) NULL);
		if (error)
			return (error);
		constty = NULL;
		return (0);
	}
	dev = cn_tab->cn_dev;
	return ((*cdevsw[major(dev)].d_ioctl)(dev, cmd, data, flag, p));
}

/*ARGSUSED*/
cnselect(dev, rw, p)
	dev_t dev;
	int rw;
	struct proc *p;
{
	if (cn_tab == NULL)
		return (1);
	return (ttselect(cn_tab->cn_dev, rw, p));
}

cngetc()
{
	if (cn_tab == NULL)
		return (0);
	return ((*cn_tab->cn_getc)(cn_tab->cn_dev));
}

cnputc(c)
	register int c;
{
	if (cn_tab == NULL)
		return;
	if (c) {
		(*cn_tab->cn_putc)(cn_tab->cn_dev, c);
		if (c == '\n')
			(*cn_tab->cn_putc)(cn_tab->cn_dev, '\r');
	}
}
